qiskit/synthesis/mcx_synthesis.py

```diff
@@ -0,0 +1,439 @@
+        q_ancillas = q[num_ctrl_qubits + 1 :]
+
+        definition = []
+
+        if self.dirty_ancillas:
+            i = num_ctrl_qubits - 3
+            ancilla_pre_rule = [
+                (HGate(), [q_target], []),
+                (CXGate(), [q_target, q_ancillas[i]], []),
+                (PhaseGate(-numpy.pi / 4), [q_ancillas[i]], []),
+                (CXGate(), [q_controls[-1], q_ancillas[i]], []),
+                (PhaseGate(numpy.pi / 4), [q_ancillas[i]], []),
+                (CXGate(), [q_target, q_ancillas[i]], []),
+                (PhaseGate(-numpy.pi / 4), [q_ancillas[i]], []),
+                (CXGate(), [q_controls[-1], q_ancillas[i]], []),
+                (PhaseGate(numpy.pi / 4), [q_ancillas[i]], []),
+            ]
+            for inst in ancilla_pre_rule:
+                definition.append(inst)
+
+            for j in reversed(range(2, num_ctrl_qubits - 1)):
+                definition.append(
+                    (RCCXGate(), [q_controls[j], q_ancillas[i - 1], q_ancillas[i]], [])
+                )
+                i -= 1
+
+        definition.append((RCCXGate(), [q_controls[0], q_controls[1], q_ancillas[0]], []))
+        i = 0
+        for j in range(2, num_ctrl_qubits - 1):
+            definition.append((RCCXGate(), [q_controls[j], q_ancillas[i], q_ancillas[i + 1]], []))
+            i += 1
+
+        if self.dirty_ancillas:
+            ancilla_post_rule = [
+                (U1Gate(-numpy.pi / 4), [q_ancillas[i]], []),
+                (CXGate(), [q_controls[-1], q_ancillas[i]], []),
+                (U1Gate(numpy.pi / 4), [q_ancillas[i]], []),
+                (CXGate(), [q_target, q_ancillas[i]], []),
+                (U1Gate(-numpy.pi / 4), [q_ancillas[i]], []),
+                (CXGate(), [q_controls[-1], q_ancillas[i]], []),
+                (U1Gate(numpy.pi / 4), [q_ancillas[i]], []),
+                (CXGate(), [q_target, q_ancillas[i]], []),
+                (U2Gate(0, numpy.pi), [q_target], []),
+            ]
+            for inst in ancilla_post_rule:
+                definition.append(inst)
+        else:
+            definition.append((CCXGate(), [q_controls[-1], q_ancillas[i], q_target], []))
+
+        for j in reversed(range(2, num_ctrl_qubits - 1)):
+            definition.append((RCCXGate(), [q_controls[j], q_ancillas[i - 1], q_ancillas[i]], []))
+            i -= 1
+        definition.append((RCCXGate(), [q_controls[0], q_controls[1], q_ancillas[i]], []))
+
+        if self.dirty_ancillas:
+            for i, j in enumerate(list(range(2, num_ctrl_qubits - 1))):
+                definition.append(
+                    (RCCXGate(), [q_controls[j], q_ancillas[i], q_ancillas[i + 1]], [])
+                )
+
+        for instr, qargs, cargs in definition:
+            qc._append(instr, qargs, cargs)
+
+        return qc
+
+
+class MCXSynthesisAlgorithms:
+    """Implements various synthesis algorithms."""
+
+    @staticmethod
+    def noancilla_3qubits(circuit_name):
```

Comment:
Wouldn't it be better to have these as separate classes implementing the `MCXSynthesis` interface? This way we can keep encapsulation (since adding a new special decomposition would not require this existing class) and we could run `isinstance` checks on them to see if they are MCX synthesis algorithms.